File: aiot_ws/src/simple_ros_cpp/src/service_server.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_srvs/srv/set_bool.hpp>
#include <chrono>
#include <iostream>

using namespace std;
using namespace std::chrono_literals;

class Service_server : public rclcpp::Node
{
public:
    Service_server()
        : Node("Service_server"){

        server_ = create_service<std_srvs::srv::SetBool>("setbool",
        std::bind(&Service_server::setBool_callback, this, std::placeholders::_1, std::placeholders::_2));

        }
private:
    // 변수형 지정 문이 너무 길어서 typedef 선언
    typedef std::shared_ptr<std_srvs::srv::SetBool::Request> BRequset;
    typedef std::shared_ptr<std_srvs::srv::SetBool::Response> BResponse;
    // self.bool = bool()
    bool _bool;
    void setBool_callback(const BRequset request, BResponse response){
        
        std::string str = request->data ? std::string("true") : std::string("false");
        RCLCPP_INFO(get_logger(), str.c_str());
        if (request->data != _bool)
        {
            _bool = !_bool;
            response->success = true;
            response->message = str + " setting success!!";
        }
        else
        {
            response->success = false;
            response->message = str + " setting fail!!";
        }
    }

    rclcpp::Service<std_srvs::srv::SetBool>::SharedPtr server_;
};

int main() {
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<Service_server>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: aiot_ws/src/simpleRos/simpleRos/service_server.py
import rclpy                    # ros2 노드 초기화, 노드구동, 노드간 통신 활성화
from time import sleep
from rclpy.node import Node     # 발행자, 구독자, 서비스, 액션, 타이머 생성
from std_srvs.srv import SetBool


class Service_server(Node):
    def __init__(self):
        super().__init__("Service_server")
        self.create_service(SetBool, "setbool", self.setBool_callback)
        self.bool = bool()

    def setBool_callback(self, request : SetBool.Request, response : SetBool.Response):
        self.get_logger().info(f"{request.data}")
        self.get_logger().info(f"{self.bool}")
        if request.data != self.bool:
            self.bool = not self.bool
            response.success = True
            response.message = f"{self.bool} setting sucess"

        else:
            response.success = False
            response.message = f"{self.bool} setting fail"
        sleep(5)
        return response


def main():
    rclpy.init()
    node = Service_server()
    
    try:
        rclpy.spin(node)
    except:
        node.destroy_node() # 키보드로 종료시 에러 비활성화

if __name__ == '__main__' :
    main()